File: kotlin/src/main/kotlin/RawSocket/protocol/RawSocketFactory.kt
package nekit.RawSocket.protocol

import nekit.Messages.ConnectSession
import nekit.Socket.SocketProtocol
import nekit.Config.NetworkInterfaceType
import nekit.Config.GlobalNetworkManager
import nekit.Utils.PlatformDetector
import nekit.RawSocket.cellular.RawCellularTCPSocket
import nekit.RawSocket.ktor.RawTCPSocket
import nekit.RawSocket.cellular.RawCellularUDPSocket
import nekit.RawSocket.ktor.RawUDPSocket

object RawSocketFactory {

    fun getRawTCPSocket(session: ConnectSession? = null): RawTCPSocketProtocol {
        val requestedInterface = if (session != null && session.interfaceType != NetworkInterfaceType.DEFAULT) {
            session.interfaceType
        } else {
            GlobalNetworkManager.activeInterface
        }

        // On Android: use RawCellularTCPSocket (supports network.bindSocket for cellular)
        // On macOS runLocal: fall back to RawTCPSocket (no cellular concept on JVM)
        return if (PlatformDetector.isAndroid && requestedInterface == NetworkInterfaceType.CELLULAR) {
            RawCellularTCPSocket()
        } else {
            RawTCPSocket()
        }
    }

    fun getRawUDPSocket(host: String, port: Int, requestedInterface: NetworkInterfaceType? = null): RawUDPSocketProtocol {
        val activeInterface = if (requestedInterface != null && requestedInterface != NetworkInterfaceType.DEFAULT) {
            requestedInterface
        } else {
            GlobalNetworkManager.activeInterface
        }

        return if (PlatformDetector.isAndroid && activeInterface == NetworkInterfaceType.CELLULAR) {
            RawCellularUDPSocket(host, port)
        } else {
            RawUDPSocket(host, port)
        }
    }
}

File: swift/src/RawSocket/RawSocketFactory.swift
import Foundation
import NetworkExtension

/**
 Represents the type of the socket.

 - NW:  The socket based on `NWTCPConnection`.
 - GCD: The socket based on `GCDAsyncSocket`.
 */
public enum SocketBaseType {
    case nw, gcd
}

/// Factory to create `RawTCPSocket` based on configuration.
open class RawSocketFactory {
    /// Current active `NETunnelProvider` which creates `NWTCPConnection` instance.
    ///
    /// - note: Must set before any connection is created if `NWTCPSocket` or `NWUDPSocket` is used.
    public static weak var TunnelProvider: NETunnelProvider?

    /**
     Return `RawTCPSocketProtocol` instance.

     - parameter type: The type of the socket (`SocketBaseType`).
     - parameter requestedInterface: The required network interface (`NetworkInterfaceType`).

     - returns: The created socket instance.
     */
    public static func getRawTCPSocket(_ type: SocketBaseType? = nil, requestedInterface: NetworkInterfaceType? = nil) -> RawTCPSocketProtocol {
        let activeInterface = requestedInterface ?? GlobalNetworkManager.shared.activeInterface

        switch type {
        case .some(.nw):
            return activeInterface == .cellular ? NWCellularTCPSocket() : NWTCPSocket()
        case .some(.gcd):
            // GCD sockets don't naturally support cellular binding through NWParameters easily here.
            return GCDTCPSocket()
        case nil:
            if RawSocketFactory.TunnelProvider == nil {
                // No tunnel extension — use NWCellularTCPSocket when cellular is requested,
                // otherwise fall back to GCDTCPSocket (OS picks the default route).
                return activeInterface == .cellular ? NWCellularTCPSocket() : GCDTCPSocket()
            } else {
                return activeInterface == .cellular ? NWCellularTCPSocket() : NWTCPSocket()
            }
        }
    }

    /**
     Return `NWUDPSocket` instance.
     Note: In Swift, GCDUDPSocket may not be fully managed by this factory, so we focus on NWUDPSocket.

     - parameter host: The host to connect to.
     - parameter port: The port to connect to.
     - parameter timeout: The timeout for the socket.
     - parameter requestedInterface: The required network interface (`NetworkInterfaceType`).

     - returns: The created socket instance or nil if initialization failed.
     */
    public static func getRawUDPSocket(host: String, port: Int, timeout: Int = Opt.UDPSocketActiveTimeout, requestedInterface: NetworkInterfaceType? = nil) -> AnyObject? {
        let activeInterface = requestedInterface ?? GlobalNetworkManager.shared.activeInterface
        
        if activeInterface == .cellular {
            return NWCellularUDPSocket(host: host, port: port, timeout: timeout)
        } else {
            return NWUDPSocket(host: host, port: port, timeout: timeout)
        }
    }
}
